File: python/백준/재귀/4779 칸토어 집합.py
def cantor_set(start, end):
    if end == 1:
        return
    
    third = end // 3
    for i in range(start + third, start + 2 * third):
        string[i] = ' '
    
    cantor_set(start, third)
    cantor_set(start + 2 * third, third)

try:
    while True:
        N = input()
        if N.strip() == "":
            break
        N = int(N)
        string = ['-'] * (3 ** N)
        cantor_set(0, 3 ** N)
        print(''.join(string))
except EOFError:
    pass

File: cpp/백준/재귀/4779 칸토어 집합.cpp
#include <iostream>
#include <string>
using namespace std;

string s;

void cantor_set(int start, int end_) {
    if (end_ == 1) return;

    int third = end_ / 3;
    for (int i = start + third; i < start + 2 * third; ++i) {
        s[i] = ' ';
    }

    cantor_set(start, third);
    cantor_set(start + 2 * third, third);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    while (cin >> N) {
        int len = 1;
        for (int i = 0; i < N; ++i) len *= 3;
        s.assign(len, '-');
        cantor_set(0, len);
        cout << s << "\n";
    }
    return 0;
}
